### PR TITLE
[shmap/partial-auto] Fixes lowering for jax.lax.axis_index in shard_map for degenerated shmaps.

### PR DESCRIPTION
**Pull request by yliu120**
Please take a look at the test I added for the example.

We found that when the manual axis has size 1, the axis_index will be definitely a replicated constant 0 across all the devices. There is no need to add iota and sharding and go full to shard for that, which actually invokes a sharding normalization inside XLA.

Let's say if we do iota and adding the sharding custom ops, then the sharding will be something like,

devices=[1,8,1]<=[8] {replicated, manual} for the iota result.

Then this sharding is normalized to {replicated} in XLA. `FullToShard` custom call cannot take `replicated` sharding which will invoke a crash in GSPMD.

### CODE REVIEW

**Comment by yliu120** [comment]
@mattjj @pschuh Can you please take a look? Thanks.

**Review by yashk2810 — APPROVED** [review]

File: tests/shard_map_test.py [hunk 1]
```diff
@@ -2162,7 +2162,22 @@ def f():
...
                        check_rep=False, auto=frozenset({'j'}))()
 
-    self.assertAllClose(f(), np.array(range(4), dtype=np.int32).reshape(-1, 1))
+    self.assertAllClose(f(), np.arange(4, dtype=np.int32).reshape(-1, 1))
+
+  def test_partial_auto_axis_index_degenerated_axis(self):
+    if config.use_shardy_partitioner.value:
+      self.skipTest('Shardy does not support full-to-shard.')
+
+    mesh = jtu.create_mesh((1, 8), ('i', 'j'))
```
Comment (yashk2810) [inline tests/shard_map_test.py:2171]: Do you mind using `(1, 2)` mesh so that we get more test coverage?

Comment (yliu120) [reply tests/shard_map_test.py]: Done.

This branch is rebased and squashed.
